kvserver,kvclient: add ExpectExclusionSince write option

stevendanna opened the pull request:
The new ExpectExclusionSince write option can use by callers of the KV API to verify that a key has not seen a write since the given timestamp. This is separate from the ReadTimestamp of the transaction because the caller may want to verify that an unreplicated exclusive lock taken at an earlier timestamp hasn't been lost.

Informs #142977

Release note: None

stevendanna committed:
kvserver,kvclient: add ExpectExclusionSince write option

The new ExpectExclusionSince write option can use by callers of the KV
API to verify that a key has not seen a write since the given
timestamp. This is separate from the ReadTimestamp of the transaction
because the caller may want to verify that an unreplicated exclusive
lock taken at an earlier timestamp hasn't been lost.

Informs #142977

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146634)
<!-- Reviewable:end -->

miraradeva reviewed (commented):
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhuRkd3Vzwp4Z1TtUQ:-OQhuRkd3Vzwp4Z1TtUR:b-evhmyo):*
nit: "can be used"
___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhuneO2vF3ShJSpNU0:-OQhuneO2vF3ShJSpNU1:b-z49ol7):*
An example would be nice here: a transaction with different `ReadTimestamp`, `WriteTimestamp`, and `ExpectExclusionSince`. 

For serializable transactions, is `ReadTimestamp` always the same as `ExpectExclusionSince`?
___
*[`pkg/kv/kvpb/api.proto` line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxKSQ9wPegynm5t9Z:-OQhxKSQ9wPegynm5t9_:bnnrdzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L298)):*
> ```proto
>   // ExpectExclusionSince, if set, indicates that the request should return an
>   // error if this key has been written to at a timestamp at or after the given
>   // timestamp. This is different from the ReadTimestamp of the request in that
> ```

The Put request doesn't really have a `ReadTimestamp`, so maybe mention that this only makes sense for transactional requests?
___
*[`pkg/kv/kvpb/api.proto` line 300 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhwJhf5eVOk-b8b2GZ:-OQhwJhf5eVOk-b8b2G_:b-o6nlyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L300)):*
> ```proto
>   // timestamp. This is different from the ReadTimestamp of the request in that
>   // it allows a writer to verify an unreplicated lock acquired at an earlier
>   // timestamp has provided the require protection.
> ```

nit: "required"
___
*[`pkg/kv/kvpb/api.proto` line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxPzw5y-BrGlwxSLP:-OQhxPzw5y-BrGlwxSLQ:btt1tqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L448)):*
> ```proto
>   // ExpectExclusionSince, if set, indicates that the request should return an
>   // error if this key has been written to at a timestamp at or after the given
>   // timestamp. This is different from the ReadTimestamp of the request in that
> ```

Same note about the `ReadTimestamp`.
___
*[`pkg/storage/mvcc.go` line 2670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhz6IAElkXRyKHTSpN:-OQhz6IAElkXRyKHTSpO:b-sfv0w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/storage/mvcc.go#L2670)):*
> ```Go
> 			writeTooOldErr := kvpb.NewWriteTooOldError(readTimestamp, writeTimestamp, key)
> 			return false, roachpb.LockAcquisition{}, writeTooOldErr
> 		} else if !exclusionTimestamp.IsEmpty() && exclusionTimestamp.LessEq(metaTimestamp) {
> ```

This is the branch of non-intent values. What happens if we want exclusion since a given timestamp and we find an intent in between `ExpectExclusionSince` and the transaction's `ReadTimestamp`? E.g. maybe we lost the unreplicated lock and an intent was written. We still have exclusion but only until that intent turns into a committed value, right? Is there anything similar we need to do for intents?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Thanks for the review!

I'll write some of the storage level test which I think will help us verify some of these cases you've asked about.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OQljbuNDvBpR6kRvaGu:b-rihctu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @miraradeva)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhuneO2vF3ShJSpNU0:-OQlg2s9AQGn2rEcQoCK:b-mbvjzc):*
> For serializable transactions, is `ReadTimestamp` always the same as `ExpectExclusionSince`?

Yes, for serialisable transactions, I think these should always be the same. This might be a nice validation to add.
___
*[`pkg/kv/kvpb/api.proto` line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxKSQ9wPegynm5t9Z:-OQlgCyT2fGUMi9MtYXR:bdcy1a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L298)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

The Put request doesn't really have a `ReadTimestamp`, so maybe mention that this only makes sense for transactional requests?
</blockquote></details>

Good point, these should all read "ReadTimestamp of the transaction" and specify that this only applies to transactional requests. This brings two additional questions I think:

1. Does this definitely work correctly during 1PC evaluation where we run the request non-transactionally.
2. What should happen for a batch that has `CanForwardReadTimestamp` set? I think in practice it should be possible to have a batch where we need this feature but where we can forward the read timestamp, perhaps that is a validation we should add.
___
*[`pkg/storage/mvcc.go` line 2670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhz6IAElkXRyKHTSpN:-OQlfixJBxPwfIAp0bEn:b-6vvmte) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/storage/mvcc.go#L2670)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This is the branch of non-intent values. What happens if we want exclusion since a given timestamp and we find an intent in between `ExpectExclusionSince` and the transaction's `ReadTimestamp`? E.g. maybe we lost the unreplicated lock and an intent was written. We still have exclusion but only until that intent turns into a committed value, right? Is there anything similar we need to do for intents?
</blockquote></details>

I'll triple check this while adding tests, but I think there are two cases:

1) The intent is from another transaction. I think in that case we should already fail in the cases above.
2) The intent is from our own transaction. I _think_ this should be allowed currently and will continue to be allowed.


<!-- Sent from Reviewable.io -->

stevendanna commented:
@arulajmani @miraradeva I have some tests to fix up, but this should be nearing its final shape. Note that a future PR would add a similar feature that would be used during replicated locking gets.

stevendanna commented on pkg/storage/testdata/mvcc_histories/del_with_exclusion_ts:
```diff
@@ -0,0 +1,78 @@
+>> at end:
+<no data>
+
+# Intent at a timestamp above the exclusion timestamp from a different transaction still results in a LockConflictError.
```
Note that this is in some sense a contention handling policy decision.  The LockConflictError doesn't escape KV and rather this will result in us waiting on transaction that slipped because of the lock loss. If it commits, we'll end up raising an exclusion violation error.

arulajmani reviewed (commented):
Mostly looks good to me, though I haven't scrutinized the test files here much -- I'll leave that to @miraradeva. 

Separately, @stevendanna and I spoke a bunch about this today, and we decided that the client side handling would work slightly differently because we're thinking of introducing a new read buffer instead of adding locks/read values to the existing b-tree. 

> Reviewed 8 of 40 files at r2, 14 of 25 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSPQsZR3TEwUaCSAw-Z:bd82hth)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @miraradeva, and @stevendanna)

___
*[`pkg/storage/mvcc.go` line 2670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhz6IAElkXRyKHTSpN:-OSPQl2J0Q2awQgFj_7_:b-3glmro) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/storage/mvcc.go#L2670)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'll triple check this while adding tests, but I think there are two cases:

1) The intent is from another transaction. I think in that case we should already fail in the cases above.
2) The intent is from our own transaction. I _think_ this should be allowed currently and will continue to be allowed.
</blockquote></details>

Agreed! 
___
*[`pkg/kv/kvpb/data.go` line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSPPgFJ5_v-zKTRSYiS:-OSPPgFJ5_v-zKTRSYiT:bd83k2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvpb/data.go#L124)):*
> ```Go
> 		// An exclusion violation error always requires a restart.
> 		txn.WriteTimestamp.Forward(tErr.RetryTimestamp())
> 		errorAlwaysRequireRestart = true
> ```

We should add a trace line here -- getting here means we must have lost a lock, which while possible, should not be common. 
___
*[`pkg/kv/kvpb/errors.proto` line 812 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSK9IMO2o6RC7K_qywc:-OSK9IMO2o6RC7K_qywd:b5xsiuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvpb/errors.proto#L812)):*
> ```proto
>   // ExclusionTimestamp is the timestamp at which the request expected its
>   // exclusive access started.
>   optional util.hlc.Timestamp exclusion_timestamp = 1 [(gogoproto.nullable) = false];
> ```

nit: s/exclusion_timesamp/expected_exclusion_since_timestamp/g
___
*[`pkg/kv/kvpb/errors.proto` line 815 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSK9Lmu-o-Irheik2qF:-OSK9Lmu-o-Irheik2qG:b-ub3p5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvpb/errors.proto#L815)):*
> ```proto
>   // ActualTimestamp is the timestamp at which the request found a write that
>   // violated its exclusive access.
>   optional util.hlc.Timestamp actual_timestamp = 2 [(gogoproto.nullable) = false];
> ```

nit: s/actual_timestamp/violation_timestamp/g
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSP1rPzDNYhDIR_d41w:-OSP1rPzDNYhDIR_d41x:bzbhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1080)):*
> ```Go
> 	// transaction, the ReadTimestamp only moves on a read refresh. If that read
> 	// refresh succeeds, then the exclusion condition must still hold and should
> 	// still hold at the time of commit (right?).
> ```

Yup, all of this checks out. If we do decide to set the exclusion timestamp just for RC/snapshot isolation transactions, then maybe we should consider having an assertion 
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSPNtMe0B-KuBlzS1NP:-OSPNtMe0B-KuBlzS1NQ:b-76cbyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1170)):*
> ```Go
> 			FoundKey: foundKey,
> 		})
> 		twb.addToBuffer(req.Key, roachpb.Value{}, req.Sequence, req.KVNemesisSeq, exclusionTS)
> ```

We spoke about this in person, but I was thinking we'd have the invariant checking when adding the exclusion timestamp to the (read) buffer. The invariant we could check is that the exclusion timestamp does not regress between each addition.
___
*[`pkg/sql/conn_executor.go` line 3589 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSP2XbN4SWHcP4cVHEg:-OSP2XbN4SWHcP4cVHEh:b-dvxkvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/sql/conn_executor.go#L3589)):*
> ```Go
> )
> 
> var allowBufferedWritesAlways = settings.RegisterBoolSetting(
> ```

nit: should we rename the variable name here, and the cluster setting, to say "weak isolation" or "read committed and snapshot" instead?

super nit: this could technically be its own commit.
___
*[`pkg/ccl/logictestccl/tests/fakedist-vec-off/BUILD.bazel` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSK9gS9-CG_zB8JDQWM:-OSK9gS9-CG_zB8JDQWN:bs8ak5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/ccl/logictestccl/tests/fakedist-vec-off/BUILD.bazel#L15)):*
> ```Starlark
>         "//conditions:default": {"test.Pool": "large"},
>     }),
>     shard_count = 35,
> ```

Confirming these are intentional?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
Looks good! I'll take another look after you make the changes from Arul's comments.

> Reviewed 15 of 40 files at r2, 25 of 25 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSUB7Ze0rKATtbxiqKv:bck6haj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxKSQ9wPegynm5t9Z:-OSQNceiCAkcI_fsKHGs:b-lgyyss) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L298)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Good point, these should all read "ReadTimestamp of the transaction" and specify that this only applies to transactional requests. This brings two additional questions I think:

1. Does this definitely work correctly during 1PC evaluation where we run the request non-transactionally.
2. What should happen for a batch that has `CanForwardReadTimestamp` set? I think in practice it should be possible to have a batch where we need this feature but where we can forward the read timestamp, perhaps that is a validation we should add.
</blockquote></details>

For 1PC, the read and write timestamps need to be the same, right? So a 1PC request shouldn't need to expect exclusion since an earlier timestamp?
___
*[`pkg/storage/mvcc.go` line 2670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhz6IAElkXRyKHTSpN:-OSUB3iXD3lvRiNoL0A3:b-46bwzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/storage/mvcc.go#L2670)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Agreed!
</blockquote></details>

Also agreed.
___
*[`pkg/kv/kvpb/errors.go` line 1896 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSQPcgU4zLPJPZNO0g0:-OSQPcgU4zLPJPZNO0g1:bxhidfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvpb/errors.go#L1896)):*
> ```Go
> 
> // RetryTimestamp returns the timestamp that should be used to retry an
> // operation after encountering a WriteTooOldError.
> ```

nit: "after encountering a ExclusionViolationError"?
___
*[`pkg/storage/testdata/mvcc_histories/put_with_exclusion_ts` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSU8sEO5paWnVx154k2:-OSU8sEO5paWnVx154k3:bvm87da) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/storage/testdata/mvcc_histories/put_with_exclusion_ts#L9)):*
> ```
> 
> run error
> put k=a v=v2 ts=3 expect-exclusion-since=1
> ```

In these tests, should we also test the case where `expect-exclusion-since` is the same as the timestamp of the previous write? In this case, 2. We still expect to see the error, right?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/buffered_writes_lock_loss` line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSQQkDz3vySqz3Bkg4H:-OSQQkDz3vySqz3Bkg4I:b-ef20l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/ccl/logictestccl/testdata/logic_test/buffered_writes_lock_loss#L40)):*
> ```
> user testuser
> 
> # Add another write to t to ensure we are propogating the ts correctly.
> ```

nit: "propagating"
___
*[`pkg/storage/mvcc.go` line 2670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSQQJ0U1vmoxIwkw_7Y:-OSQQJ0U1vmoxIwkw_7Z:b-q5mj4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/storage/mvcc.go#L2670)):*
> ```Go
> 			return false, roachpb.LockAcquisition{}, writeTooOldErr
> 		} else if !exclusionTimestamp.IsEmpty() && exclusionTimestamp.LessEq(metaTimestamp) {
> 			// TODO(ssd): Does the timestamp check above need to be inclusive or exclusive?
> ```

Inclusive maybe. We have a locking read evaluated at that timestamp, right?


<!-- Sent from Reviewable.io -->

stevendanna committed:
kvcoord: track min sequence number for locks in txnWriteBuffer

To correctly handle the expected exclusion timestamp with rollbacks,
we need to track the sequence number at which our write inclusion was
established as well as the timestamp that it was established at.

Informs #142977

Release note: None

stevendanna commented on pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:
```diff
@@ -1394,6 +1427,14 @@ type bufferedWrite struct {
+	// TODO(review): Are we OK with requiring exclusion across savepoint
+	// rollbacks? If not, we'll need to track this on a per-value level and then
+	// and flush time send the lowest value when writing in the key. That's not hard
+	// but it isn't clear to me what behavior we expect.
```
What I've done as a first step is move this into a lockKeyInfo struct that tracks the sequence numbers and allows us to handle rollback correctly.  I've left it here for now because this PR is already getting rather large. I'll tackle potentially moving this to its own b-tree in the PR that decomposes unreplicated locks.
<!-- Reviewable comment -OSUG13pB-VMXSDlIZSy:bboo64n -->
<!-- Sent from Reviewable.io -->

miraradeva replied:
To make sure I understand the rollback issue: if we didn't track these locks and required exclusion from the lowest timestamp, we would get unnecessary retries but we won't violate correctness? It's too aggressive to expect such exclusion and it makes no sense to do it for rolled back writes?
<!-- Reviewable comment -OSZI_bA35k724PIrB4G:b-qupbyf -->
<!-- Sent from Reviewable.io -->

stevendanna replied:
> It's too aggressive to expect such exclusion and it makes no sense to do it for rolled back writes?

Yes, I think you are correct. To be honest the rollback stuff might be a _bit_ overkill. But, my thinking was that if the write is rolled back, then we either (1) won't have another write on that key and don't need the exclusion timestamp tracked anymore or (2) will have a write on this key which will likely issue a new locking request. If we didn't drop the locks and always sent the lowest timestamp we ever saw, we could perhaps elide a few more locking requests. But then whether you would get an erroneous retry would depend on whether the lock table actually rolled back your lock. Now, the _current_ code here is complicated enough to read that it felt like just assuming it can is the best bet.

This is also a _little_ forward looking since we will be able to elide a broader set of locking requests if we know what locks have already been obtained.
<!-- Reviewable comment -OSagcD09czPOYmt2pRU:b8rys17 -->
<!-- Sent from Reviewable.io -->

miraradeva replied:
Makes sense. Now that you mentioned it, I thought the lock table doesn't release locks on savepoint rollback? Did we change this recently?
<!-- Reviewable comment -OSdL1eC-xNFTs1vXh-o:b-2t6fr0 -->
<!-- Sent from Reviewable.io -->

stevendanna replied:
I re-read through this this morning in response to your comment and I couldn't find a way to make it happen even though I had convinced myself this was required a few days before.
<!-- Reviewable comment -OSd_7u5DnkuM99WNy_V:b8g1xt0 -->
<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
TFTR!

Handling rollbacks correctly here was unfortunately a good bit of code. Most of it will be required for buffering replicated locking requests anyway. But, I think we should aim for getting some portion of this in to avoid this becoming a 2k line PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSVMoaq5lEqom1jCYms:b-hhanio)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @miraradeva)

___
*[`pkg/ccl/logictestccl/tests/fakedist-vec-off/BUILD.bazel` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSK9gS9-CG_zB8JDQWM:-OSUIqIBBoAFbN6OhLqy:b-jmvelf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/ccl/logictestccl/tests/fakedist-vec-off/BUILD.bazel#L15)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Confirming these are intentional?
</blockquote></details>

They are the result ./dev generate. I'll confirm before merge.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSP1rPzDNYhDIR_d41w:-OSUIxoQ1lkmTuq4iHaq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1080)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Yup, all of this checks out. If we do decide to set the exclusion timestamp just for RC/snapshot isolation transactions, then maybe we should consider having an assertion
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSPNtMe0B-KuBlzS1NP:-OSUFuHA9aZvpNsDdQzF:b-mcha7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1170)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We spoke about this in person, but I was thinking we'd have the invariant checking when adding the exclusion timestamp to the (read) buffer. The invariant we could check is that the exclusion timestamp does not regress between each addition.
</blockquote></details>

I've added this to the current code for now as I think it'll be cleaner to do the larger rework we talked about in another PR.
___
*[`pkg/kv/kvpb/api.proto` line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxKSQ9wPegynm5t9Z:-OSUH7beCAo5a4Rflb0Q:b-k7bm5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L298)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

For 1PC, the read and write timestamps need to be the same, right? So a 1PC request shouldn't need to expect exclusion since an earlier timestamp?
</blockquote></details>

It's true that they shouldn't need the exclusion timestamp. But, it may be set on the request anyway. That  should still be OK.
___
*[`pkg/kv/kvpb/api.proto` line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhxPzw5y-BrGlwxSLP:-OSUHgtS4cBGDG8Ae8Ts:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/kv/kvpb/api.proto#L448)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Same note about the `ReadTimestamp`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/data.go` line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSPPgFJ5_v-zKTRSYiS:-OSUIhd97GmcqOTPdA4-:b-iedsp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/kv/kvpb/data.go#L124)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We should add a trace line here -- getting here means we must have lost a lock, which while possible, should not be common.
</blockquote></details>

I _think_ think should already end up in the trace in a few places. I'll double chekc out.
___
*[`pkg/storage/mvcc.go` line 2670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OQhz6IAElkXRyKHTSpN:-OSVL_o33gAETY1LcvmU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f469299a24dabb3263866efc5c5fab616505a4d/pkg/storage/mvcc.go#L2670)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Also agreed.
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 2670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSQQJ0U1vmoxIwkw_7Y:-OSUJXAf96qGL-RDjqt6:b-8ayu6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/storage/mvcc.go#L2670)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Inclusive maybe. We have a locking read evaluated at that timestamp, right?
</blockquote></details>

Agree
___
*[`pkg/storage/testdata/mvcc_histories/put_with_exclusion_ts` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSU8sEO5paWnVx154k2:-OSVLkJXDcfBdv0GCKOB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a9d35e9a917ffbfa21519cc219405ca96b02cfd/pkg/storage/testdata/mvcc_histories/put_with_exclusion_ts#L9)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

In these tests, should we also test the case where `expect-exclusion-since` is the same as the timestamp of the previous write? In this case, 2. We still expect to see the error, right?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
Looks good overall. My main comment is around the complexity of handling the locking strengths in such a general way, given that we only care about two (exclusive and shared).

Speaking of which, I'm not 100% sure why we care about shared locks in the write buffer. Buffered-writes decomposition generates only exclusive locks; we need to handle them correctly and ensure exclusion. It seems like we have shared locks here just for the cases where a user explicitly acquires an unreplicated shared lock? Don't we still treat these as best effort for serializable, and we acquire them as replicated locks for read committed?

> Reviewed 11 of 11 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSZSOZJEbB15ylQqK4N:b-nsk3r3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2428 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZQqnK0ifgwFWDKWj8:-OSZQqnK0ifgwFWDKWj9:b-bnwylj) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2428)):*
> ```Go
> func TestLockKeyInfo(t *testing.T) {
> 	ts1 := hlc.Timestamp{WallTime: 1}
> 	ts2 := hlc.Timestamp{WallTime: 1}
> ```

Are these timestamps intentionally the same? I guess they don't really matter.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1471 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZL_MU68U2szV9RRDe:-OSZL_MU68U2szV9RRDf:bhm9524) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1471)):*
> ```Go
> }
> 
> func (bw *bufferedWrite) acquireLock(li *lockAcquisition) {
> ```

`newLockedKeyInfo` seems to do something very similar to this function. Can we use it here? In either case, it seems like this logic belongs to a `lockKeyInfo` function, to ensure the correct timestamp manipulation.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1610 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZL-RP1KjFACZ5zVb1:-OSZL-RP1KjFACZ5zVb2:bq7kdh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1610)):*
> ```Go
> 	// strength was acquired.
> 	//
> 	// This read should be removed from the tree when it isn't held at any level.
> ```

nit: This sentence seems a bit out of place. We haven't really talked about a read, though I imagine it's the locking read what acquired the lock. We also haven't mentioned a tree yet.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1611 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZKeF940e73paKKae2:-OSZKeF940e73paKKae3:b-cjwwqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1611)):*
> ```Go
> 	//
> 	// This read should be removed from the tree when it isn't held at any level.
> 	heldStrengths [2]enginepb.TxnSeq
> ```

Since it's just two strengths, maybe it's less error-prone to have two separate fields? Do we plan to track more in this slice in the future?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1698 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZPhWK5js560NbM9j9:-OSZPhWK5js560NbM9jA:blpgvs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1698)):*
> ```Go
> // greater.
> func (li *lockedKeyInfo) heldGE(str lock.Strength) bool {
> 	// heldStrengths
> ```

nit: remove?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> I'm not 100% sure why we care about shared locks in the write buffer.

Yes, handling Shared locks here is forward looking to the next PR when we start transforming replicated locking requests into unreplicated locking requests. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSacKra3ifGO6uScyZE:b-r2ffl1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @miraradeva)



<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @miraradeva)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1471 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZL_MU68U2szV9RRDe:-OSanVwxBlW1KDBKIcKe:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1471)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

`newLockedKeyInfo` seems to do something very similar to this function. Can we use it here? In either case, it seems like this logic belongs to a `lockKeyInfo` function, to ensure the correct timestamp manipulation.
</blockquote></details>

Good call. Done.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1611 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZKeF940e73paKKae2:-OSangX-4YqZK3IUf9Lh:bqqtbsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1611)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Since it's just two strengths, maybe it's less error-prone to have two separate fields? Do we plan to track more in this slice in the future?
</blockquote></details>

:D I keep going back and forth on this.  I'll address your other comments and then play around with this locally to see which I like better. The downside 
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2428 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZQqnK0ifgwFWDKWj8:-OSaoP80-HFAIUqZd-IZ:b-rlbxjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2428)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Are these timestamps intentionally the same? I guess they don't really matter.
</blockquote></details>

Oversight. Update this and added a couple of assertions that depends on it.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-:-OSdLXR83iqbV-sy6nSH:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1611 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/146634#-OSZKeF940e73paKKae2:-OSdLNCkAdk94ZUw5p6r:bpm8fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/71316ac50a64204b18d02ae1e52e5847bd0915ed/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1611)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

:D I keep going back and forth on this.  I'll address your other comments and then play around with this locally to see which I like better. The downside
</blockquote></details>

Yeah, up to you.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146634)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @miraradeva)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=miraradeva
